src/animation.rs

```diff
@@ -0,0 +1,151 @@
+#![allow(dead_code)]
+
+use cgmath;
+use cgmath::prelude::*;
+use mint;
+
+use Object;
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum LoopKind {
+    Once,
+    Repeat,
+    PingPong,
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Interpolation {
+    Discrete,
+    Linear,
+    Smooth,
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Ending {
+    ZeroCurvature,
+    ZeroSlope,
+    WrapAround,
+}
+
+#[derive(Clone, Debug)]
+pub struct Loop {
+    clip: Clip,
+    time: f32,
+}
+
+impl Loop {
+    pub fn new(clip: Clip) -> Self {
+        Loop {
+            time: 0.0,
+            clip,
+        }
+    }
+
+    pub fn update(&mut self, delta_time: f32) {
+        self.time += delta_time;
+        for mut track in &mut self.clip.tracks {
+            match track.target {
+                keyframe::Target::Orientation(ref frames) => {
+                    let orientation = frames.interpolate(self.time);
+                    track.object.set_orientation(orientation);
+                },
+                keyframe::Target::Position(ref frames) => {
+                    let position = frames.interpolate(self.time);
+                    track.object.set_position(position);
+
+                },
+                keyframe::Target::Scale(ref frames) => {
+                    let scale = frames.interpolate(self.time);
+                    track.object.set_scale(scale);
+                },
+            }
+        }
+    }
+}
+
+/// A reusable set of keyframe tracks which represent an animation.
+#[derive(Clone, Debug)]
+pub struct Clip {
+    /// A name for this clip.
+    pub name: Option<String>,
+
+    /// The animation keyframe tracks.
+    pub tracks: Vec<keyframe::Track>,
+}
+
+pub mod keyframe {
+    use cgmath;
+    use mint;
+    use super::Interpolation;
+    use Object;
+
+    /// The frames of a target.
+    #[derive(Clone, Debug)]
+    pub struct Frames<T> {
```

Comment:
I'm trying to mimic the [`times` / `values`](https://threejs.org/docs/#api/animation/KeyframeTrack.times) properties in a type-safe way here.